lib/mirage.ml

```diff
@@ -541,80 +538,107 @@ let ip = Type IP
 
-let ipv4_conf ?address ?netmask ?gateways () = impl @@ object
+let ipv4_keyed_conf ?address ?network ?gateway () = impl @@ object
     inherit base_configurable
     method ty = ethernet @-> arpv4 @-> ipv4
     method name = Name.create "ipv4" ~prefix:"ipv4"
-    method module_name = "Ipv4.Make"
+    method module_name = "Static_ipv4.Make"
     method packages = Key.pure ["tcpip"]
     method libraries = Key.pure ["tcpip.ipv4"]
-    method keys = address @?? netmask @?? gateways @?? []
+    method keys = address @?? network @?? gateway @?? []
     method connect _ modname = function
       | [ etif ; arp ] ->
         Fmt.strf
           "%s.connect@[@ %a@ %a@ %a@ %s@ %s@]"
           modname
           (opt_key "ip") address
-          (opt_key "netmask") netmask
-          (opt_key "gateways") gateways
+          (opt_key "network") network
+          (opt_key "gateway") gateway
           etif arp
       | _ -> failwith "The ipv4 connect should receive exactly two arguments."
   end
 
-let create_ipv4
-    ?(clock = default_monotonic_clock) ?(time = default_time)
-    ?group net { address ; netmask ; gateways } =
-  let etif = etif net in
-  let arp = arp ~clock ~time etif in
-  let address = Key.V4.ip ?group address in
-  let netmask = Key.V4.netmask ?group netmask in
-  let gateways = Key.V4.gateways ?group gateways in
-  ipv4_conf ~address ~netmask ~gateways () $ etif $ arp
+let dhcp_conf = impl @@ object
+  inherit base_configurable
+  method ty = time @-> network @-> dhcp
+  method name = "dhcp"
+  method module_name = "Dhcp_client_mirage.Make"
+  method packages = Key.pure [ "charrua-client" ]
+  method libraries = Key.pure [ "charrua-client.mirage" ]
+  method connect _ modname = function
+  | [ _time; network ] ->
+    Fmt.strf
+      "%s.connect %s "
+      modname network
+  | _ -> failwith "The dhcp_config connect should receive exactly two arguments."
+end
+
+let ipv4_dhcp_conf = impl @@ object
+    inherit base_configurable
+    method ty = dhcp @-> ethernet @-> arpv4 @-> ipv4
+    method name = Name.create "ipv4" ~prefix:"ipv4"
+    method module_name = "Dhcp_ipv4.Make"
+    method packages = Key.pure ["tcpip"]
+    method libraries = Key.pure ["tcpip.ipv4"]
+    method connect _ modname = function
+          | [ dhcp ; ethernet ; arp ] ->
+        Fmt.strf
+          "%s.connect@[@ %s@ %s@ %s@]"
+          modname
+          dhcp ethernet arp
+      | _ -> failwith "The ipv4 connect should receive exactly three arguments."
+  end
 
-let default_ipv4_conf =
-  let i = Ipaddr.V4.of_string_exn in
-  {
-    address  = i "10.0.0.2";
-    netmask  = i "255.255.255.0";
-    gateways = [i "10.0.0.1"];
-  }
 
-let default_ipv4 ?group net = create_ipv4 ?group net default_ipv4_conf
+let dhcp time net = dhcp_conf $ time $ net
+let ipv4_of_dhcp dhcp ethif arp = ipv4_dhcp_conf $ dhcp $ ethif $ arp
 
-type ipv6_config = (Ipaddr.V6.t, Ipaddr.V6.Prefix.t list) ip_config
+let create_ipv4
+    ?group _net etif arp =
```

Comment:
Since this is breaking compat, why not drop the `net` argument (and take a `etif` in `create_ipv6`) ?